pkg/grafana/client.go

```diff
@@ -0,0 +1,137 @@
+package grafana
+
+import (
+	"errors"
+	"net/url"
+	"strings"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/kanopy-platform/k8s-auth-portal/pkg/random"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	ErrRoleNotValid = errors.New("role is not valid")
+)
+
+type Client struct {
+	client GAPIClient
+}
+
+type GAPIClient interface {
+	UserByEmail(email string) (user gapi.User, err error)
+	CreateUser(user gapi.User) (int64, error)
+	AddOrgUser(orgID int64, user, role string) error
+	UpdateOrgUser(orgID, userID int64, role string) error
+}
+
+func NewClient(baseURL *url.URL, cfg gapi.Config) (*Client, error) {
+	newClient := &Client{}
+
+	client, err := gapi.New(baseURL.String(), cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	newClient.client = client
+
+	return newClient, nil
+}
+
+// LookupUser search for a user by Login or Email and returns it
+func (c *Client) LookupUser(loginOrEmail string) (*gapi.User, error) {
+	user, err := c.client.UserByEmail(loginOrEmail)
+	if err != nil {
+		if !strings.Contains(err.Error(), "User not found") {
+			return nil, err
+		}
+	}
+
+	// gapi returns an empty struct
+	if user.Login == "" {
+		return nil, nil
+	}
+
+	return &user, nil
+}
+
+// CreateUser adds a new global user to Grafana
+func (c *Client) CreateUser(user gapi.User) (int64, error) {
```

Comment:
The reason I did that was to have LookupUser return a nil if a user is not present, but it can be an empty user so I agreed that working with objects instead of references everywhere makes more sense

Reply:
I agree, but after seeing it in use and the disconnect it creates with the other functions, it loses that nil check benefit, since you already need to test for an "empty" object in other areas.

Reply:
agreed, updated all methods to conform to the same api